# How can MiGA help me?

MiGA's analyses can illuminate community composition, identify taxonomic
novelty, hypothesize gene prediction, and evaluate assembly quality. MiGA can
take in a variety of inputs such as raw, unassembled reads, assembled isolate
genomes, metagenome-assembled genome (MAGs) and single-cell amplified genomes
(SAGs). MiGA uses a combination of the genome-aggregate average nucleotide
identity concept or ANI and the average amino-acid identity, AAI, to
taxonomically classify a query genomic sequence against the genome sequences in
its reference database. Part of MiGAâ€™s strength lies in the 10,000 reference
genomes that make up its database and an efficient heuristic algorithm to search
the query genome against all these genomes. The reference database is regularly
updated and improved with minimal downtime. This ensures consistently improved
accuracy of classification without a large impact to its users.

